quadratichq/quadratic
# feat: better zero states around code

## Overview

Introduce placeholder text for the code editor that provides better heuristics for learning how the mechanics of the app work.

<img width="527" alt="CleanShot 2023-05-30 at 16 43 08@2x" src="https://github.com/quadratichq/quadratic/assets/1316441/0e6328a4-5388-4fbc-8925-b2bfaa5197dc">

- Placeholder text disappears as you type (empty value brings it back).
- Placeholder text has links to auto-fill the code editor with useful snippets.
- Placeholder text has preference to never display that message again

The console now also has placeholder text:

<img width="521" alt="CleanShot 2023-05-31 at 16 54 59@2x" src="https://github.com/quadratichq/quadratic/assets/1316441/b7631519-9cdf-4258-898b-b7b4ec77a05e">

And when you're doing formulas, the `print` suggestion is gone:

<img width="459" alt="CleanShot 2023-05-31 at 16 55 33@2x" src="https://github.com/quadratichq/quadratic/assets/1316441/e26b212c-36b5-4788-bbdd-8106174076d1">


### To Test

- Only works for Python (no placeholder text for AI/Formulas)
- Code snippets generate code that work immediately upon being generated
  - The "cell reference" is an exception, as it generates `cell(x,y)` because we don't know where valid on the sheet is (and we don't want to generate an infinite loop)
- User preference to hide placeholder shows up in localstorage (click "don't show again" then clear in localstorage to ensure it comes back)

### Notes

Prior art: VSCode's "New file" placeholder text

<img width="723" alt="CleanShot 2023-05-26 at 15 47 04@2x" src="https://github.com/quadratichq/quadratic/assets/1316441/18821058-10ad-4569-8632-79439bcbf269">
